Java/core/src/main/java/com/amazon/randomcutforest/util/ArrayPacking.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.util;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+
+import java.util.Arrays;
+
+public class ArrayPacking {
+
+    static int maxPack(int base) {
+        int pack = 0;
+        long num = base;
+        while (num < Integer.MAX_VALUE) {
+            num = num * base;
+            ++pack;
+        }
+        return pack;
+    }
+
+    public static int[] compress(short[] inputArray, boolean compress) {
+        if (inputArray == null) {
+            throw new IllegalStateException(" array packing invoked on null arrays");
+        }
+        if (!compress || inputArray.length < 3) {
```

Comment:
I think it should be an exception if `compress` is true and `inputArray.length < 3`.

Reply:
Well, every array will fail to give any benefit for some length. A tree of size 2 can be defined (of a large size tree is defined and only only two leaves are present for now).